### ISSUE TITLE
Improve post reading, video and picture viewing for users with screen readers

### ISSUE THREAD

**Issue by juanreina921** [QuantumBadger/RedReader#799]
Hey! I tried the app out, and it's grate! My name is Juan, and I'm a blind user of android. This app does everything the official Reddit app doesn't. Almost everything from menus and options, links, buttons, checkboxes, and the usual UI element's have lables on them, and are detected by the screen reader. The posts can be long pressed to show the menu, the interface is all round good, except for some issues which I will cover. 
When ever I double tap to go and read the reddit post, my screen reader doesn't report anything, regardless if I swipe up down, left or rite, as if there was nothing on the page. Double tapping the screen returns you back to the previous window with everything being interactable again. However, if I long pressed to bring up the menu, I can view comments to the post, but I still can't view the first post at all, only the comments. I tried playing with the various settings to no avail. 
Another issue, is video and image viewing. When I double tap on a reddit post that has a picture or video, the screen reader reports loading, but like the previous issue, the first post is not readable, and double tapping the screen jumps you back to the previous window. I tried to get the controlls to show as well as playing once more with the settings again, only to no avail. 
I reinstalled the app, but still the above issues persisted. I would appreciate if these issues could be resolved, as the app is already 99 accessible, it just has little more to go. I'd love to give feedback and help if you like.

**Comment by QuantumBadger** [comment]
Thanks for the suggestions! When I get a free moment I'll look into this.

**Comment by juanreina921** [comment]
@QuantumBadger, thank you so much! If you'd like more ideas I'm glad to help. Also sorry for closing and reopening the issue, github on Android isn't as clear when labelling some of these buttons.

**Comment by QuantumBadger** [comment]
@juanreina921 I've installed TalkBack and tried to reproduce these issues. However it works correctly for me -- I'm able to read reddit posts after double tapping.

Could you provide more details please about what you're trying to do, for example are you trying to view: an image post, a web link post, or a comment section?

If it's possible to send a video of what's happening (including the screen reader audio) that would be very helpful. Also, please could you confirm which screen reader you're using?

**Comment by juanreina921** [comment]
Hello there. @QuantumBadger, I did a recording of talkback and my primary screen reader commentary screen reader. Ps. I'm very sorry about the audio and its quality, it's not the best.
Talkback link= []()http://www.fileconvoy.com/dfl.php?id=g5f915fd6880c2e151000298829a3ad3b5728145701
Commentary link= []()http://www.fileconvoy.com/dfl.php?id=g45922b0f5adcc889100029883024baba2802c4cb38
The files will be there for three weeks, so don't worry. I also tried to discuss as clearly as I could. If you need jieshuo/commentary, let me know. Remember, not all screen readers are the same, so it is good to have another to test with. I look forward to your ideas, Juan.

**Comment by QuantumBadger** [comment]
Thanks for the recordings! No need to apologise for the audio quality, I appreciate the effort.

I've made some changes to the app based on your comments:

* When viewing an image, the title of the Activity is announced as "Viewing image".
* There is now a hidden menu in the image viewer, which only screen readers can see. This currently has the options "view comments" and "go back".
* The subtitles under posts (containing the number of points, author, etc), and the headers over comments, didn't sound great, so I've changed this to be clearer to screen readers. For example, instead of saying "123 p t s 3 hours ago by QuantumBadger" it will now say: "Score: 123. Age: 3 hours ago. Author: QuantumBadger". This is a bit slower to read out but I think it's clearer.

One suggestion I haven't done yet is adding media controls (like pause/etc) to videos. Are the existing controls accessible or would it be better to have them in the new hidden menu as well?

I've tried the changes with TalkBack, if you'd like to try it yourself and give me some feedback I've put a link to Alpha 243 below:

https://redreader.org/alpha/builds/RedReader_Alpha_243.apk

**Comment by juanreina921** [comment]
Hey there. @QuantumBadger, I've tested the alpha out, and you've outdone yourself!  I like what you did, the subtitle announces flare, domain, author, score, and age. Don't worry about slowness of speach, alot of blind people have trained ourselves on understanding fast speach, I had to take a normal speaking rate just in case you couldn't understand. With video viewing, in alpha, the video controls work, with one button. in the hidden menu, on commentary there's only one unlabeled button, that button plays and pauses video playback. Don't worry, all that button needs is a label. Might I suggest adding a mute button, and a slider to change video position? But awesome work man! Those are my ideas for now, I look forward to hearing yours!

**Comment by codeofdusk** [comment]
Wow, this is a really great app and might become an equivalent to [Dystopia](https://reddit.com/r/dystopiaforreddit) for Android with just a bit of work, which would be huge for blind users like me!

A few asks:
* There's no need to include the words "score", "age", "flair", or "domain" in the accessible label as it only slows down the experience for screen reader users as these words don't really give any useful info. Maybe put the word "points" or "up" after the score?
* Please add custom accessibility actions for up- and down-voting comments and submissions.
* In the accessibility label of each comment, specify its relative level in the thread. For instance, responses to OP are at level 1, children of responses to OP are l2, and so on.
* Bonus: add accessibility actions to "collapse" comments of a given level. In other words, allow the user to move easily to the next sibling or cousin by way of two separate accessibility actions. Perhaps call the first "collapse comment" and the second "collapse thread".

A good accessible Reddit client would allow me to more fully use Android and is one of the major things keeping me on iOS, so this would definitely mean a lot to me and other blind Android users!

**Comment by codeofdusk** [comment]
CC @Cguy7777.

**Comment by juanreina921** [comment]
@codeofdusk, the reason this is done is to be clear about what the numbers ment, instead of having them in a big jumble. I think that you can change the titles like score, age, etc in settings/behavior.  Personally, I think it's much clearer this way, especially if you're speech rate is fast, so you don't have slowdown. Hope this helps.
